### DISCUSSION TITLE
Theme designer loads cached theme.designer.json even after Theme change

### DISCUSSION THREAD

**Discussion by X3Technology** [vanjarosoftware/Vanjaro.Platform#1700] [General]
I made a copy of the Basic theme.  I then edited the new theme.designer.json and stripped out everything except the Site->Colors controls.
![just_colors](https://user-images.githubusercontent.com/1585717/131217235-660e8dc9-7051-4367-aa35-936199cec568.PNG)

This works as expected and only shows the controls left in my new Theme. (yes, I know this breaks a whole bunch of Vanjaro blocks like Login and such, but I'm just messing around).

I then switched the Site theme back to the Basic theme, but the designer did not switch back to the collection on controls for Basic, it still loaded the same collection for my new theme.  I suspect the theme.designer.json file is cached after switching themes.  The Basic themes controls would not load until after I clicked the Save Theme button on the UX.

This is what I expected to see after Applying the Basic theme:

![basic_controls](https://user-images.githubusercontent.com/1585717/131217433-7ac63b01-3b67-4453-a9a6-1903ff5543b0.PNG)

The UX controls should switch automatically whenever switching themes.

**Comment by meetmandeep** [comment]
> I made a copy of the Basic theme. 

You cannot simply copy a theme folder and create a copy. Themes have a server side component. We will provide official documentation on how to create new theme starting next month dev series webinars.

**Comment by X3Technology** [comment]
Maybe you are referring to the theme.css file as the server component?  I can confirm that making a copy of the Basic folder does in fact allow you to have a custom theme, with its own set of customable Theme Editor controls.

Maybe its not supposed to work, but it does.

**Comment by mksmanish79** [comment]
You cannot simply copy a theme folder and create a copy. Its showing you in the themes list as theme works as a combination of folder and server side components. 

The issue we are able to duplicate is editing any theme is editing first theme. We will fix this in next update.

**Comment by X3Technology** [comment]
Can you please explain what you mean by 'server side components'?  Because following the steps I did sure does seem to have the exact same effect of setting up a new theme.  

I am able to switch between the two themes.  I am able to add my own custom editor controls to my custom theme, edit the default Sass, and remove some controls that I didn't want.  I can switch back and forth between the two themes and see different sets of editor controls being loaded and different style choices being applied to the page.

So, even if I have set up a new theme "incorrectly", the end effect is that everything works as expected.  So, what am I missing?

The only additional step that is required is that you have to recycle the app pool to clear the cache when switching themes as it appears the theme.editor.json file is cached.  Restarting the application after applying a new theme fixes it.  Perhaps the 'server side component' you are referring to prevents this caching problem?

**Reply by mksmanish79** [reply]
You need to implement ITheme interface to your theme so that you can Import/Export your and settings are updated based on itheme name.

https://github.com/vanjarosoftware/Vanjaro.Platform/blob/develop/DesktopModules/Vanjaro/Core/Library/Components/Interfaces/ITheme.cs

"The only additional step that is required is that you have to recycle the app pool to clear the cache when switching themes as it appears the theme.editor.json file is cached."

This is a bug and will be fixed in next update. To make it work correctly make sure you only edit the theme by applying the theme as settings are binding based on current applied theme on the portal.

**Comment by X3Technology** [comment]
Ah!  OK!  I think I actually did that.  In the theme folder, I noticed the theme.cs file.  I edited this file and changed the name property to my theme name and also put in a unique GUID.

Is that all that is necessary?

**Reply by mksmanish79** [reply]
What you are doing is fine for now but its not recommended way of copying a theme. Unique Theme GUID is required for future functionality. Theme.cs should not be there as it will be compiled in dll.
